node/hyperbahn/service_proxy.js

```diff
@@ -295,11 +299,60 @@ ServiceDispatchHandler.prototype.refreshServicePeer =
 function refreshServicePeer(serviceName, hostPort) {
     var self = this;
 
+    // Add hostPort to service peers, decision pending on whether to connect.
     var peer = self.getServicePeer(serviceName, hostPort);
-    peer.connect();
 
+    // The old way: fully connect every egress to all affine peers.
+    if (!self.partialAffinityEnabled) {
+        peer.connect();
+        return;
+    }
+
+    // Make a note not to collect this service
     var time = self.channel.timers.now();
     self.exitServices[serviceName] = time;
+
+    // Obtain and sort the affine worker and relay lists.
+    var relays = Object.keys(self.egressNodes.exitsFor(serviceName));
+    relays.sort();
+    var serviceChannel = self.getOrCreateServiceChannel(serviceName);
+    var workers = serviceChannel.peers.keys();
+    workers.sort();
+
+    // Find our position within the affine relay set so we can project that
+    // position into the affine worker set.
+    var relayIndex = relays.indexOf(self.channel.hostPort);
+    if (relayIndex < 0) {
+        // This should only occur if an advertisement loses the race with a
+        // relay ring membership change.
+        self.logger.warn('Relay could not find itself in the affinity set for service', {
+            serviceName: serviceName,
+            relayHostPort: self.channel.hostPort,
+            workerHostPort: hostPort,
+            relays: relays,
+            workers: workers
+        });
+        return;
+    }
+
+    // Compute the range of workers that this relay should be connected to.
+    var ratio = workers.length / relays.length;
+    var start = (relayIndex * ratio) % workers.length;
+    var length = Math.max(self.minPeersPerRelay, self.minPeersPerWorker * ratio);
+    var stop = (start + length) % workers.length;
+
+    // Adjust connections as necessary.
+    for (var workerIndex = 0; workerIndex < workers.length; workerIndex++) {
```

Comment:
Please don't do a o(workers) loop with a branch in it; instead please have a separate passage for "connect to these", and separate the disconnect / reaping concern.  +:100: if we find a way to do that disconnect in less than o(workers).

I'm imagining something like:
```
for (var i = start, stop = start + length; i < stop; i++) {
  connectTo(workers[i % workers.length]);
  update(someFastDataStructureToMakeTheReapingNotSuck);
}
```